pkg/xds/envoy/types.go

```diff
@@ -195,6 +195,19 @@ func (sa ServicesAccumulator) Services() Services {
 	return sa.services
 }
 
+// Clusters returns a slice of all clusters gathered in all accumulated
+// services. This is a convenience method, which may be used whenever there is
+// need to configure tcpProxy with weighted clusters (like in MeshTCPProxy)
+func (sa ServicesAccumulator) Clusters() []Cluster {
```

Comment:
Where is this needed exactly, that we can "throw away" the information which service it belongs to? I.e. why not do `sa.Services()` and `range` over that?

Reply:
here: https://github.com/kumahq/kuma/pull/6873/files#diff-62202617ea4d191debba0e1bdc0150b4b4dc8d1c5f61ac54bac945664134cb66R14

Reply:
Should it maybe be in #6911? Why don't you need to use the return value of `MakeHTTPSplit` like https://github.com/kumahq/kuma/pull/6873/files#r1211377734?

Reply:
`tcpProxy` expects full `Cluster` structures, and splits doesn't contain `tags` if we would like to change it (ref. 
- https://github.com/kumahq/kuma/blob/bb8688bb01e338436c2b21bf6522309150858702/pkg/plugins/policies/xds/split.go#L3-L8
- https://github.com/kumahq/kuma/blob/52045f6f74d789e390d1a3e54c0ff2bfc80b8b8d/pkg/xds/envoy/listeners/v3/tcp_proxy_configurer.go#L50

Reply:
It would be similarly used as in:
https://github.com/kumahq/kuma/blob/6c0e32225279ad6186f105719fd8aeaaaba6d782/pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go#L64-L68
I don't like it to be honest (implicit modifications of ServiceAccumulator inside this function, but I didn't have enough time to try to change it)

Reply:
It looks like we don't actually use that Metadata when we call `TcpProxy()`, only calls to `TcpProxyWithMetadata` do, which is apparently only used for egress/ingress proxies.

Reply:
Yes, we could try to change the `TcpProxyConfigurer` to accept some common for Clusters and Splits abstraction, but it would involve:
- Changing the name of the method for one of these to have consistent `Name` or `ClusterName`, where `Name` doesn't exactly fit to `Split` and `ClusterName` doesn't exactly fit to `Cluster`
- Refactor `TcpProxyConfigurer` to remove constrain of having to provide structure with `Tags` inside
Then still it wouldn't be fully correct as if there are multiple splits to the same clusters, I think they should be summed up in a one `WeightedCluster` (I don't even know if envoy would allow to point to clusters with the same name inside tcp proxy filter)

Reply:
Actually, I just checked and I can't find any logic in envoy's code base, which would prevent in providing multiple clusters with the same name inside tcp proxy filter (ref. https://github.com/envoyproxy/envoy/blob/ba9a2a7b0ee700117a49f493ff72fe080c5c2b9e/source/common/tcp_proxy/tcp_proxy.cc#L47), so we probably could use Splits without worrying of this, but this needs to be empirically confirmed.

Reply:
As an option we can simply add tags to split:

```go
type Split struct {
	clusterName string
	weight      uint32
	lbMetadata  tags.Tags

	hasExternalService bool
}
```

It seems like tags make sense both for Split and Cluster so I think it's fine to have them in 2 places. In that case we can make `TcpProxyConfigurer` to use only Splits and avoid depending on Clusters at all